### DISCUSSION TITLE
getting 504 from latest version but not from 0.9

### DISCUSSION THREAD

**Discussion by Fredrikp-ume** [JuliaWeb/HTTP.jl#924] [General]
Hi, I have successfully used HTTP.jl for a microservice and it has been great. 

When migrating to version newer than 1.0 from 0.9 I'm getting 504 from my React client for even the simplest requests that has previously worked.

I've tried with both julia 1.7.3 and 1.8.0 with the same result

anyone come across this problem?

best wishes
Fredrik

**Comment by quinnj** [comment]
I haven't seen any issues like this. If you can share an example of what exactly you're calling like `HTTP.request("GET", ...; verbose=2)` and share the output, that would help a lot in seeing what exactly is going on under the hood and what might have changed.

Just as a wild guess, a 504 is usually a gateway timeout and can crop up when you get redirected to a service that is unavailable. We did slightly change the redirect logic to follow the official RFC more correctly. I know one other user had to pass `redirect_method=:same` to get a certain scenario to work.

**Reply by Fredrikp-ume** [reply]
Thanks for your quick reply, appreciated!

I've borrowed some patterns from your awesome MusicAlbums example ;) 

Even from a first login I'm getting a 504 that is not happening with 0.9.17
**Frontend**
```export function loginToBackend(credentials = {}, onLogin){
  return fetch(`/api/user/login`, {
    method:"POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(credentials),
  }).then((response) => {

    let token = response.headers.get('Jwt-Token')

    if(!token){
      throw new TypeError("Oops, the login failed! Wrong credentials or user does not exist in backend");      
    }

    onLogin({"Jwt-Token":token})
  }).catch(error => console.error(error));
};
```
**HTTP.jl**
```
const ROUTER = HTTP.Router()

function contextHandler(req)

    user = Auth.User(req)

    withcontext(user) do
        HTTP.Response(200, JSON3.write(ROUTER(req)))
    end
end

const AUTH_ROUTER = HTTP.Router(contextHandler)

function authenticate(user::User)

    resp = HTTP.Response(200, JSON3.write(user))
    return Auth.addtoken!(resp, user)
end

loginUser(req) = authenticate(Service.loginUser(JSON3.read(req.body, User))::User)
HTTP.register!(AUTH_ROUTER, "POST", "/user/login", loginUser)
```

For this endpoint it works and still returns the token but for other endpoints it doesn't work at all.
[
![image](https://user-images.githubusercontent.com/24318111/189680358-bb27896b-e375-402f-85b0-b766d4dc0843.png)
](url)

**Reply by Fredrikp-ume** [reply]
... and it work flawlessly with Postman or curl

**Reply by Fredrikp-ume** [reply]
[
![image](https://user-images.githubusercontent.com/24318111/189687689-f94eb504-8d12-4ecc-8195-20be3cdb2071.png)
](url)

**Reply by quinnj** [reply]
Hmmmm, it's a little hard to tell exactly what's going on from the limited context (if you have the full code somewhere, I'm happy to take a look). But one thing I noticed is that the `AUTH_ROUTER` should be defined like:
```julia
const AUTH_ROUTER = HTTP.Router(HTTP.Handlers.default404, HTTP.Handlers.default405, contextHandler)
```
If you're passing a middleware to the router. As the 1st argument, `contextHandler` would be used to respond to 404s.

**Reply by Fredrikp-ume** [reply]
Thanks! The application is quite complex so I'll try to strip it down enough to still give the same error.

**Reply by Fredrikp-ume** [reply]
Have cleaned up the application so it essentially mainly contains some boilerplate from the MusicAlbums example ;)

 https://github.com/Fredrikp-ume/httpjl_debug

both endpoints, test and login, gives 504

test occasionally works using fetch in React

both always work with Postman

with HTTP.jl 0.9.17 everything works

**Reply by Fredrikp-ume** [reply]
Hi @quinnj ,
just checking if you have had time to look at the issue?

It works if I would downgrade so it is no major problem. Mainly curious ;)
